dev-docs/RFCs/v8.0/carto-v9-rfc.md

```diff
@@ -0,0 +1,222 @@
+
+## API background
+
+Without getting into the details too much, using the current `CartoLayer` with the CARTO works in the following way:
+
+1. User configures `CartoLayer` with a set of props
+2. `CartoLayer` uses these props to construct a URL for a Map Instantiation request
+3. `CartoLayer` performs a Map Instantiation fetch request and obtains URLs for data endpoints, in different formats: TileJSON, GeoJSON, JSON etc
+4. `CartoLayer` is [hardcoded](https://github.com/visgl/deck.gl/blob/master/modules/carto/src/layers/carto-layer.ts#L263) to only use TileJSON, so the TileJSON URL is fetched
+5. `CartoLayer` [decides](https://github.com/visgl/deck.gl/blob/master/modules/carto/src/api/layer-map.ts#L203-L223) which Layer to use for rendering and passes the TileJSON payload to that layer in the `data` prop
+
+## Proposal
+
+At a high level we want to achieve an API which makes it clearer to the user what is going on, and allows for better extensibility. We also want to separate the options used for loading the data from the props which style the layer.
+
+To do so, the interaction with the CARTO API will be done via a set of typed helper classes which aid the user in obtaining the TileJSON needed for a particular layer.
+
+```javascript
+import CartoQuerySource, CartoQuerySourceOptions, CartoTileLayer from '@deck.gl/carto';
+
+// Global options
+setDefaultCredentials({accessToken: 'XXXX'});
+
+// All properties in one place
+const options: CartoQuerySourceOptions = {
+  connectionName: 'bigquery',
+  sqlQuery: 'select * from carto.tables.example'
+}
+
+// Returns a Promise which resolves to TileJSON payload. Can optionally await if needed
+const data = CartoQuerySource(options);
+
+const layer = new CartoTileLayer({
+  data,
+
+  // ...style props
+});
+```
+
+## Class names
+
+The goal is to have descriptive names for classes and options, even if this means they are more verbose.
+
+| Source class                | Layer                   |
+| --------------------------- | ----------------------- |
+| `CartoTableSource`          | `CartoTileLayer`        |
+| `CartoTilesetSource`        | `CartoTileLayer`        |
+| `CartoQuerySource`          | `CartoTileLayer`        |
+| `CartoH3TableSource`        | `CartoH3TileLayer`      |
+| `CartoH3QuerySource`        | `CartoH3TileLayer`      |
+| `CartoH3TilesetSource`      | `CartoH3TileLayer`      |
+| `CartoQuadbinTableSource`   | `CartoQuadbinTileLayer` |
+| `CartoQuadbinQuerySource`   | `CartoQuadbinTileLayer` |
+| `CartoQuadbinTilesetSource` | `CartoQuadbinTileLayer` |
+| `CartoRasterSource`         | `CartoRasterTileLayer`  |
+
+### Options
+
+Each of the `XXXSource` classes takes an options object as a constructor parameter, all inheriting from the common `CartoSourceOptions` base type.
+
+```javascript
+type CartoSourceOptions = {
```

Comment:
I would include the path URL of maps. We could potentially change the path parameter of maps api and for widgets we will need the path to maps api.

mapsApiUrlPath = /v3/maps
(future) sqlApiUrlPath = /v3/sql
(future) statsApiUrlPath = /v3/stats